README.md
# ESP Web Config

A Platform IO library to configure ESP32 & ESP8266 devices over a web interface.

## ❓ How it works

- Sets WiFi to Access point.
- Runs a web server of a static web app from SPIFFS (ESP32) or LittleFS (ESP8266) with default 170Kb size.
- Uses web basic authentication (username/password).
- Accepts configuration data via a personalized form.
- Saves configuration to Non-volatile Storage via `Preferences` library.

## Web application

The served web app is located under `app` folder.
It's built with [Vue.js 3](https://vuejs.org/). Uses [Tailwindcss](https://tailwindcss.com/) for styling and [FormKit](https://formkit.com/) for dynamic form generation. No need to be concerned about the tech stack, we just need the bundle located under `data` folder.

## Form generation

The form definition is decoupled from the web application. It's defined by `schema.json` located under `data` folder. The schema is a JSON object interpreted by [Formkit](https://formkit.com/essentials/schema). It supports styling, validation, conditional rendering and more.

## ESP8266 support

To use this library with `espressif8266` platform, you should

- Install [vshymanskyy/Preferences](https://registry.platformio.org/libraries/vshymanskyy/Preferences) library.
- Specify `LittleFS` as the file system.

```ini
# ./platformio.ini

board_build.filesystem = littlefs
```

## Setup

- Download `data` folder from this [link](https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/becem-gharbi/esp-web-config/tree/main/data) and place it in your project root.
- Edit `schema.json` file depending on your use case.
- Build Filesystem image.
- Upload Filesystem image.

## Default configuration

| Parameter  | Value          |
| ---------- | -------------- |
| host       | 192.168.4.1:80 |
| ssid       | MAC address    |
| passphrase | 123456789      |
| username   | admin          |
| password   | 1234           |

## Usage

The library exposes `WebConfig` class which has the following methods:

| Method | Description                                  |
| ------ | -------------------------------------------- |
| init   | Initialize Non-volatile storage              |
| read   | Get a configuration field                    |
| write  | Set a configuration field                    |
| begin  | Start the web server                         |
| end    | Stop the web server and restart the software |
| reset  | Clear the configuration                      |

## License

[MIT License](./LICENSE)


CHANGELOG.md
# Changelog


## v0.2.1

[compare changes](https://github.com/becem-gharbi/esp-web-config/compare/v0.1.2...v0.2.1)


### 🚀 Enhancements

  - Add support for ESP8266 ([23cc708](https://github.com/becem-gharbi/esp-web-config/commit/23cc708))

### 🩹 Fixes

  - End server on config set terminated ([121cbe1](https://github.com/becem-gharbi/esp-web-config/commit/121cbe1))

### 💅 Refactors

  - Add _registerHandlers method ([1b07e75](https://github.com/becem-gharbi/esp-web-config/commit/1b07e75))

### 📖 Documentation

  - Update readme ([5fcfe9e](https://github.com/becem-gharbi/esp-web-config/commit/5fcfe9e))
  - Add ESP8266 support section to readme ([848a53e](https://github.com/becem-gharbi/esp-web-config/commit/848a53e))

### 🏡 Chore

  - Set github release ([1afca99](https://github.com/becem-gharbi/esp-web-config/commit/1afca99))
  - Set esp8266 and esp32 env in platformio.ini ([7c5f223](https://github.com/becem-gharbi/esp-web-config/commit/7c5f223))
  - Add ESP8266 support to library.json ([9c5c362](https://github.com/becem-gharbi/esp-web-config/commit/9c5c362))

### ❤️  Contributors

- Becem-gharbi

## v0.1.2


### 🚀 Enhancements

  - Add web server ([6763f4f](https://github.com/becem-gharbi/esp-web-config/commit/6763f4f))
  - Add vue app and serve it from SPIFFS ([971c6e8](https://github.com/becem-gharbi/esp-web-config/commit/971c6e8))
  - End server & SPIFFS on end method ([83a309c](https://github.com/becem-gharbi/esp-web-config/commit/83a309c))
  - Add key/value storage via Preference library ([ca89b3d](https://github.com/becem-gharbi/esp-web-config/commit/ca89b3d))
  - Add config POST server handler ([992cf78](https://github.com/becem-gharbi/esp-web-config/commit/992cf78))
  - Add config/set server handler ([3df1050](https://github.com/becem-gharbi/esp-web-config/commit/3df1050))
  - Add basic auth on config handlers ([557f258](https://github.com/becem-gharbi/esp-web-config/commit/557f258))
  - Add schema loader and interpreter ([9b5476a](https://github.com/becem-gharbi/esp-web-config/commit/9b5476a))
  - Add form submission handler ([015babb](https://github.com/becem-gharbi/esp-web-config/commit/015babb))

### 🩹 Fixes

  - Remove git submodules ([9f232e0](https://github.com/becem-gharbi/esp-web-config/commit/9f232e0))
  - Ckeck if key exists on read ([1bb22bb](https://github.com/becem-gharbi/esp-web-config/commit/1bb22bb))
  - Fix init handling ([493b0d3](https://github.com/becem-gharbi/esp-web-config/commit/493b0d3))
  - Fix form value binding ([e9f8977](https://github.com/becem-gharbi/esp-web-config/commit/e9f8977))

### 💅 Refactors

  - Set basic code structure ([db63c03](https://github.com/becem-gharbi/esp-web-config/commit/db63c03))
  - Change begin type to bool ([2b904a1](https://github.com/becem-gharbi/esp-web-config/commit/2b904a1))
  - Request authentication on page load ([d4d439c](https://github.com/becem-gharbi/esp-web-config/commit/d4d439c))
  - Perform software reset on end ([0ca8396](https://github.com/becem-gharbi/esp-web-config/commit/0ca8396))

### 📖 Documentation

  - Update readme ([069852c](https://github.com/becem-gharbi/esp-web-config/commit/069852c))
  - Update readme ([2c50dce](https://github.com/becem-gharbi/esp-web-config/commit/2c50dce))
  - Update readme ([bff1e68](https://github.com/becem-gharbi/esp-web-config/commit/bff1e68))
  - Update readme ([5175df5](https://github.com/becem-gharbi/esp-web-config/commit/5175df5))

### 🏡 Chore

  - Add manifest file ([9d62deb](https://github.com/becem-gharbi/esp-web-config/commit/9d62deb))
  - Set initial library version to 0.0.1 ([926354e](https://github.com/becem-gharbi/esp-web-config/commit/926354e))
  - Add basic example ([147fbb3](https://github.com/becem-gharbi/esp-web-config/commit/147fbb3))
  - Rename readme ([57aee36](https://github.com/becem-gharbi/esp-web-config/commit/57aee36))
  - Change library version to 0.1.0 ([b4f4336](https://github.com/becem-gharbi/esp-web-config/commit/b4f4336))
  - Add changelogen ([e2087ef](https://github.com/becem-gharbi/esp-web-config/commit/e2087ef))

### ❤️  Contributors

- Becem-gharbi

## v0.1.1


### 🚀 Enhancements

  - Add web server ([6763f4f](https://github.com/becem-gharbi/esp-web-config/commit/6763f4f))
  - Add vue app and serve it from SPIFFS ([971c6e8](https://github.com/becem-gharbi/esp-web-config/commit/971c6e8))
  - End server & SPIFFS on end method ([83a309c](https://github.com/becem-gharbi/esp-web-config/commit/83a309c))
  - Add key/value storage via Preference library ([ca89b3d](https://github.com/becem-gharbi/esp-web-config/commit/ca89b3d))
  - Add config POST server handler ([992cf78](https://github.com/becem-gharbi/esp-web-config/commit/992cf78))
  - Add config/set server handler ([3df1050](https://github.com/becem-gharbi/esp-web-config/commit/3df1050))
  - Add basic auth on config handlers ([557f258](https://github.com/becem-gharbi/esp-web-config/commit/557f258))
  - Add schema loader and interpreter ([9b5476a](https://github.com/becem-gharbi/esp-web-config/commit/9b5476a))
  - Add form submission handler ([015babb](https://github.com/becem-gharbi/esp-web-config/commit/015babb))

### 🩹 Fixes

  - Remove git submodules ([9f232e0](https://github.com/becem-gharbi/esp-web-config/commit/9f232e0))
  - Ckeck if key exists on read ([1bb22bb](https://github.com/becem-gharbi/esp-web-config/commit/1bb22bb))
  - Fix init handling ([493b0d3](https://github.com/becem-gharbi/esp-web-config/commit/493b0d3))
  - Fix form value binding ([e9f8977](https://github.com/becem-gharbi/esp-web-config/commit/e9f8977))

### 💅 Refactors

  - Set basic code structure ([db63c03](https://github.com/becem-gharbi/esp-web-config/commit/db63c03))
  - Change begin type to bool ([2b904a1](https://github.com/becem-gharbi/esp-web-config/commit/2b904a1))
  - Request authentication on page load ([d4d439c](https://github.com/becem-gharbi/esp-web-config/commit/d4d439c))
  - Perform software reset on end ([0ca8396](https://github.com/becem-gharbi/esp-web-config/commit/0ca8396))

### 📖 Documentation


src/WebConfig.cpp
#include "WebConfig.h"

WebConfig::WebConfig(int port = 80) : _server(port)
{
    _initialized = false;
}

void WebConfig::init()
{
    if (_initialized)
    {
        return;
    }

    _prefs.begin("web-config", false);

    if (read("username").length() == 0)
    {
        write("username", "admin");
    }

    if (read("password").length() == 0)
    {
        write("password", "1234");
    }

    if (read("ssid").length() == 0)
    {
        write("ssid", WiFi.macAddress());
    }

    if (read("passphrase").length() == 0)
    {
        write("passphrase", "123456789");
    }

    _initialized = true;
}

bool WebConfig::begin()
{
    init();

    // Initialize Filesystem

#ifdef ESP8266
    if (!LittleFS.begin())
    {
        Serial.println("[WebConfig] Failed to start LittleFS");
        return false;
    }
#else
    if (!SPIFFS.begin(true))
    {
        Serial.println("[WebConfig] Failed to start SPIFFS");
        return false;
    }
#endif

    // Register server handlers
    _registerHandlers();

    // Setup an AP
    if (!WiFi.softAP(read("ssid"), read("passphrase")))
    {
        Serial.println("[WebConfig] Failed to start AP");
        return false;
    }

    // Start server
    _server.begin();

    Serial.print("[WebConfig] Started on ");
    Serial.println(WiFi.softAPIP());

    return true;
}

void WebConfig::_registerHandlers()
{
    _server.on("/app.css", HTTP_GET, [](AsyncWebServerRequest *request)
               {
#ifdef ESP8266
                   request->send(LittleFS, "/app.css", "text/css");
#else
            request->send(SPIFFS, "/app.css", "text/css");
#endif
               });

    _server.on("/app.js", HTTP_GET, [](AsyncWebServerRequest *request)
               {
#ifdef ESP8266
                   request->send(LittleFS, "/app.js", "text/javascript");
#else
            request->send(SPIFFS, "/app.js", "text/javascript");
#endif
               });

    _server.on("/favicon.svg", HTTP_GET, [](AsyncWebServerRequest *request)
               {
#ifdef ESP8266
                   request->send(LittleFS, "/favicon.svg", "image/svg+xml");
#else
            request->send(SPIFFS, "/favicon.svg", "image/svg+xml");
#endif
               });

    _server.on("/schema.json", HTTP_GET, [](AsyncWebServerRequest *request)
               {
#ifdef ESP8266
                   request->send(LittleFS, "/schema.json", "application/json");
#else
            request->send(SPIFFS, "/schema.json", "application/json");
#endif
               });

    _server.on("/", HTTP_GET, [&](AsyncWebServerRequest *request)
               {
                   if (!request->authenticate(read("username").c_str(), read("password").c_str()))
                   {
                       return request->requestAuthentication();
                   }
#ifdef ESP8266
                   request->send(LittleFS, "/index.html", "text/html");
#else
            request->send(SPIFFS, "/index.html", "text/html");
#endif
               });

    _server.addHandler(
        new AsyncCallbackJsonWebHandler(
            "/config/set",
            [&](AsyncWebServerRequest *request, JsonVariant &jsonVariant)
            {
                if (!request->authenticate(read("username").c_str(), read("password").c_str()))
                {
                    return request->requestAuthentication();
                }

                JsonObject jsonObj = jsonVariant.as<JsonObject>();

                for (JsonPair p : jsonObj)
                {
                    String key = p.key().c_str();
                    String value = p.value();
                    write(key, value);
                }

                request->send(200, "text/plain", "OK");

                request->onDisconnect([&]()
                                      { end(); });
            }));

    _server.addHandler(
        new AsyncCallbackJsonWebHandler(
            "/config/get",
            [&](AsyncWebServerRequest *request, JsonVariant &jsonVariant)
            {
                if (!request->authenticate(read("username").c_str(), read("password").c_str()))
                {
                    return request->requestAuthentication();
                }

                JsonObject inputObj = jsonVariant.as<JsonObject>();

                DynamicJsonDocument doc(1024);
                JsonObject outputObj = doc.to<JsonObject>();

                for (JsonPair p : inputObj)
                {
                    String key = p.key().c_str();
                    outputObj[key] = read(key);
                }

                AsyncResponseStream *response = request->beginResponseStream("application/json");

                serializeJson(outputObj, *response);
                request->send(response);
            }));
}

void WebConfig::end()
{

#ifdef ESP8266
    LittleFS.end();
#else
    SPIFFS.end();
#endif

    _server.end();
    ESP.restart();
}

String WebConfig::read(String key)
{
    const char *_key = key.c_str();
    if (_prefs.isKey(_key))
    {
        return _prefs.getString(_key);
    }
    return "";
}

void WebConfig::write(String key, String value)
{
    _prefs.putString(key.c_str(), value);
}

bool WebConfig::reset()
{
    return _prefs.clear();
}
